### PR TITLE
Optimize ranged_balance.cpp

### PR DESCRIPTION
**Pull request by SlackingSleeper**
#### Summary

SUMMARY: Infrastructure "Makes tests go vrooom. Makes it easier to read and understand the code."
#### Purpose of change

To make the code more readable to new contributors and to make the tests faster.

When I first encountered the code it was barely readable. Calling a function with `range_test( {{ -1, -1, 0.1, -1, -1 }} )` says nothing about what you're putting into the function.

#### Describe the solution

Replaced the previous system with the Threshold class.
Added multithreading to `fire_test` to make it faster.

This does not alter the results of the tests.

### CODE REVIEW

File: tests/ranged_balance.cpp [hunk 1]
```diff
@@ -95,38 +120,78 @@ static void equip_shooter( npc &shooter, const std::vector<std::string> &apparel
...
+    for( size_t i = 0; i < sample_size; i++ ) {
+        const projectile_attack_aim aim = projectile_attack_roll( dispersion, range, 0.5 );
+        firing_stats.add( aim.missed_by < threshold.accuracy() );
+    }
+    mtx.lock();
+    feed_me.add( firing_stats );
+    mtx.unlock();
+}
+
+
```
Comment (Night-Pryanik) [inline tests/ranged_balance.cpp]:
```suggestion
```

**Comment by kevingranade** [comment]
I'm deeply opposed to adding multithreading anywhere unless absolutely necessary.

The test suite is immensely, embarrassingly parallelizeable. Just partition into equally sized sub-suites and launch them in parallell, this can be done in the makefile.

**Comment by SlackingSleeper** [comment]
> 
> 
> I'm deeply opposed to adding multithreading anywhere unless absolutely necessary.
> 

Could you go into detail why?

> The test suite is immensely, embarrassingly parallelizeable. Just partition into equally sized sub-suites and launch them in parallell, this can be done in the makefile.

I'll look into it.

**Comment by kevingranade** [comment]
In broad strokes:
It adds a great deal of cognitive load to developers to reason about in-process concurrent accesses to data.
To meaningfully leverage multi-threading, the entire codebase has to be overhauled to provide concurrency guarantees.
Bugs caused by in-process concurrency are extremely difficult to debug.
Locking overhead and thread context switching makes multithreading slower on single-core systems, the slowest systems we support *are* single-core, so it makes the already marginal systems worse in order to make already ok systems faster.
In the context of DDA, there are much simpler ways to obtain most of the speedups we care about.

Concerning how to parallellize the test suite:
In the makefile, add one invocation of the test executable per concurrency level, pass the concurrency level and the number of the invocation as arguments.
i.e. cata_test --concurrency 5 -invocation_number 1, cata_test --concurrency 5 --invocation_number 2, etc
In test main, enumerate the tests to be invoked and partition them evenly based on the passed in concurrency level.
Execute the ith partition locally and then exit.
Also in the makefile, capture the output of each invocation to a temp file, and once tests are complete, cat the temp files to the terminal.

**Comment by SlackingSleeper** [comment]
The problem with parallellizing the tests is that synthesized_range_test takes 10 times longer than the rest of the tests combined - with the quickdraw section taking the majority of the duration. Even if I were to split the sections into seperate parts the slowest one would take 5 times as long as the current test suite.

Running cata_tests without arguments takes ~150 seconds on my rig. Synthetic_range_test takes ~1600s on master, ~600s with multithreading.

What are the first two sections used for? Could an alternative be to remove them?

The code should only run the multithread part on systems with multiple cores. The current implementation would make it slightly slower on single core systems due to an added if-statement but that should be able to be fixed without too much of an hazzle.

**Comment by kevingranade** [comment]
Why is this even an issue? This is a test/analysis fixture, but not something that is run automatically, so it doesn't really matter how long it takes to run, it certainly doesn't justify adding bespoke multithreading support to speed it up by 3x. I'm absolutely certain you can speed it up more than that by simply tweaking the implementation, it's utterly unoptimized since nothing about it requires high execution speed.

Regardless, multithreading is an absolute last resort, I'm serious when I say this is a *hard* no.

If you need to just run the "good hit thresholds" part by itself, you can, see https://github.com/catchorg/Catch2/blob/master/docs/command-line.md#specify-the-section-to-run

tl;dr ``cata_test "synthetic_range_test" --section "good hit thresholds"``

**Comment by SlackingSleeper** [comment]
I see. Thanks for the advice.

Would it be better to close this PR and resubmit the code cleaning as its own PR or remove the multithreading from this one?

**Comment by kevingranade** [comment]
Updating this PR is fine, if you don't squash away the commits adding threading I will.

**Comment by SlackingSleeper** [comment]
I'll do it - it'll be good practice. :)
